[New Entry] C++: Overloading (Operators & Functions)

KTom101 opened the pull request:
<!--
👋 Hi, thanks for sending a PR to Codecademy Docs! 💖
Please fill out all fields below to ensure your PR is reviewed quickly.

**IMPORTANT**

If you would like to receive credit for your contribution to an entry, make sure to have your Codecademy user profile linked to your GitHub account:

1. Go to your Codecademy dashboard.
2. Click on your profile image in the top-right and then choose "Profile".
3. Then click "Edit Profile".
4. In the "GitHub Username" field, add your username (without the @).
5. Click "Save Changes"

Of course, you can opt not to do this and be listed as an "Anonymous contributor", instead. :)
-->

### Description

New entry on overloading operators and functions in C++; backlog from keyword shortlist. Please let me know if the file placement is fine; would we possibly want this under Operators? I'm open to changes. 

<!-- Please write a summary of the change, such as which topic(s) and file(s) that you have edited or created. Please also include relevant motivation and context: -->

### Type of Change

<!--- Please delete or cross off the bullet point(s) that are irrelevant to this PR: -->

- Adding a new entry

### Checklist

<!-- Please check ALL the boxes: -->

- [x] All writings are my own.
- [x] My entry follows the Codecademy Docs style guide.
- [x] My changes generate no new warnings.
- [x] I have performed a self-review of my own writing and code.
- [x] I have checked my entry and corrected any misspellings.
- [x] I have made corresponding changes to the documentation if needed.
- [x] I have confirmed my changes are not being pushed from my forked `main` branch.
- [x] I have confirmed that I'm pushing from a new branch named after the changes I'm making.
- [x] Under "Development" on the right, I have linked any issues that are relevant to this PR (write "Closes #<issue number> in the "Description" above).

<!--
Having trouble with the PR checker? Here are some common issues and resolutions:

- verify_formatting is failing
  - run `yarn format path/to/markdown/file.md` or `yarn format:all` and commit the results
- verify_lint is failing
  - same as above
  - if verify_lint is still failing, running `yarn lint` locally should let you know what needs to be changed by hand
- test is failing
  - ensure any new markdown files have a `Title` and `Description` defined in their metadata
  - ensure any new markdown files only contain alphanumerics and dashes in their file names and have the same name as their parent directory
  - if that looks ok, running `yarn test` locally should let you know what the issue is
-->

SSwiniarski commented on content/cpp/concepts/operator-overloading/operator-overloading.md:
```diff
@@ -0,0 +1,111 @@
+
+**Operator overloading** redefines operators for user-defined classes. When an overloaded operator is called, the compiler determines which definition to use based on the arguments provided. The following operators can be overloaded:
+
+| ->      | ->\* | new    | new [] | delete | delete []     |
```
The table is sort of cryptic. Can we have some description on what the layout, columns and rows, actually means?

KTom101 replied:
@SSwiniarski These are all the operators that could be overloaded. I was going to list them but thought it would be too long so that's why I put it in a table. What do you suggest?

SSwiniarski replied:
@KTom101 A bulleted list would be better.

SSwiniarski commented on content/cpp/concepts/operator-overloading/operator-overloading.md:
```diff
@@ -0,0 +1,111 @@
+
+The operators below cannot be overloaded:
+
+| ::  | .\* | .   | ?   |
```
Why is this a table rather than a bulleted list?

KTom101 replied:
@SSwiniarski Should I change both tables to a bulleted list?

SSwiniarski replied:
@KTom101 I think it would be less confusing.

SSwiniarski commented on content/cpp/concepts/operator-overloading/operator-overloading.md:
```diff
@@ -0,0 +1,111 @@
+    }
+
+    // Overload + operator
+    Pyramid operator+(const Pyramid& p){
```
Is there supposed to be no space between the operator and everything else? There were spaces in the syntax example.

KTom101 replied:
@SSwiniarski I'm not entirely sure what the typical convention is for this. It works with the spacing too. I was taught the way I wrote it here in the example, with no spacing. It is also written with no spacing in Bjarne Stroustrop's "The C++ Programming Language". I do think the pseudo syntax is easier to read with spacing though. Let me know what you think.

SSwiniarski replied:
@KTom101 I agree that the syntax parses better with the spaces. As long as it works with the spaces as well, then I don't think there's a real issue.

SSwiniarski reviewed (changes requested):
@KTom101 I did a first review and left a couple of comments. I see arguments for and against placing this as a term under operators, but I'd like to see what other opinions are: @Dusch4593, @yangc95?

Dusch4593 commented:
> @KTom101 I did a first review and left a couple of comments. I see arguments for and against placing this as a term under operators, but I'd like to see what other opinions are: @Dusch4593, @yangc95?

My recommendation would be to repurpose as a concept entry simply entitled "Overloading", that includes information about how it is applied to [operators](https://www.codecademy.com/resources/docs/cpp/operators) and [functions](https://www.codecademy.com/resources/docs/cpp/functions) in C++ (with links to those existing concepts included).

KTom101 commented:
> > @KTom101 I did a first review and left a couple of comments. I see arguments for and against placing this as a term under operators, but I'd like to see what other opinions are: @Dusch4593, @yangc95?
> 
> My recommendation would be to repurpose this as a concept entry simply entitled "Overloading", that includes information about how it is applied to [operators](https://www.codecademy.com/resources/docs/cpp/operators) and [functions](https://www.codecademy.com/resources/docs/cpp/functions) in C++ (with links to those existing concepts included).

@Dusch4593 To clarify, we're keeping this where it is now (as a concept entry) and just renaming it so it can be for both operators and functions?

Dusch4593 commented:
> @Dusch4593 To clarify, we're keeping this where it is now (as a concept entry) and just renaming it so it can be for both operators and functions?

@KTom101 Exactly! 😄 This would be renamed as "Overloading" and remain as a concept entry.

KTom101 commented:
@SSwiniarski @Dusch4593  I've added changes including changing the title and file name to "Overloading". I've added the topic of function overloading and also changed the table of operators to an unordered list format. 

Please let me know what additional changes should be made.

Dusch4593 commented on content/cpp/concepts/overloading/overloading.md:
```diff
@@ -0,0 +1,199 @@
+#include <iostream>
+using namespace std;
+
+class Pyramid{
+  public:
+  // Defining getters and setters
+    double getVolume(void){
+      return (base*height)/3;
+    }
+
+    void setBase(double b){
+      base = b;
+    }
+
+    void setHeight(double h){
+      height = h;
+    }
+
+    // Overload + operator
+    Pyramid operator+(const Pyramid& p){
+      Pyramid pyramid;
+      pyramid.base = this->base + p.base;
+      pyramid.height = this->height + p.height;
+      return pyramid;
+    }
```
```suggestion
class Pyramid {
  public:
  // Defining getters and setters
    double getVolume(void) {
      return (base*height)/3;
    }

    void setBase(double b) {
      base = b;
    }

    void setHeight(double h) {
      height = h;
    }

    // Overload + operator
    Pyramid operator + (const Pyramid& p) {
      Pyramid pyramid;
      pyramid.base = this->base + p.base;
      pyramid.height = this->height + p.height;
      return pyramid;
    }
    
```

Dusch4593 commented on content/cpp/concepts/overloading/overloading.md:
```diff
@@ -0,0 +1,199 @@
+      pyramid.height = this->height + p.height;
+      return pyramid;
+    }
+  private:
+    double base;
+    double height;
+};
+
+int main(){
+  // Declare pyramids and volume variables
+  Pyramid pyramid1;
+  Pyramid pyramid2;
+  Pyramid pyramid3;
+  double vol = 0.0;
+
+  // Setting pyramid1
+  pyramid1.setBase(22.0);
+  pyramid1.setHeight(16.0);
+  vol = pyramid1.getVolume();
+  cout << "Volume of pyramid1 is: " << vol << endl;
+
+  // Setting pyramid2
+  pyramid2.setBase(18.0);
+  pyramid2.setHeight(14.0);
+  vol = pyramid2.getVolume();
+  cout << "Volume of pyramid2 is: " << vol << endl;
```
```suggestion
  private:
    double base;
    double height;
};

int main() {
  // Declare pyramids and volume variables
  Pyramid pyramid1;
  Pyramid pyramid2;
  Pyramid pyramid3;
  double vol = 0.0;

  // Setting pyramid1
  pyramid1.setBase(22.0);
  pyramid1.setHeight(16.0);
  vol = pyramid1.getVolume();
  cout << "Volume of pyramid1 is: " << vol << endl;

  // Setting pyramid2
  pyramid2.setBase(18.0);
  pyramid2.setHeight(14.0);
  vol = pyramid2.getVolume();
  cout << "Volume of pyramid2 is: " << vol << endl;
```

Dusch4593 reviewed (dismissed):
@KTom101 Left some comments/suggestions for you. Thanks for the work put into this, so far. 😄

KTom101 commented on content/cpp/concepts/overloading/overloading.md:
```diff
@@ -0,0 +1,199 @@
+## Operator Overloading
+
+Operator overloading redefines built-in operators for user-defined classes. When an overloaded operator is called, the compiler determines which definition to use based on the arguments provided. The following operators can be overloaded:
+
```
@Dusch4593 I do like this better. 

@SSwiniarski Does a table like this work for you?

KTom101 commented on content/cpp/concepts/overloading/overloading.md:
````diff
@@ -0,0 +1,199 @@
+Since a function can be overloaded multiple times, the syntax can look different case-to-case. The following is an outline of what an overloaded function could look like:
+
+```pseudo
+functionName(parameters1);
+functionName(parameters2);
+...
````
@Dusch4593 I just had it to indicate there can be more than two definitions. But I will take it out!

Dusch4593 replied:
@KTom101 Gotcha! 😄

In that case, should we mention this somewhere below line # 32?

SSwiniarski commented:
@KTom101, let me know when you're done with @Dusch4593's feedback, then I'll review this again.

KTom101 commented:
@SSwiniarski I've added the changes, let me know what you think. Also, perhaps this could use one more review afterward.

SSwiniarski reviewed (approved):
@KTom101 Looks good for another review.

yangc95 reviewed (approved):
Made a minor change, but otherwise looks good to merge!

KTom101 commented:
Live Link: https://www.codecademy.com/resources/docs/cpp/overloading

cc @yangc95 @SSwiniarski
